File: Assets/EchoSim/Simulation/Goals/GoalModel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using EchoSim.Core;

namespace EchoSim.Simulation
{
    /// <summary>Additive personality contribution: Min + (Max-Min)*traitValue.</summary>
    public sealed class TraitTerm
    {
        public PersonalityTrait Trait { get; }
        public float Min { get; }
        public float Max { get; }

        public TraitTerm(PersonalityTrait trait, float min, float max)
        {
            Trait = trait;
            Min = min;
            Max = max;
            if (min > max) throw new ArgumentException("Min must be <= Max.");
        }
    }

    /// <summary>Need-driven term: curve(need/100) * weight.</summary>
    public sealed class NeedTerm
    {
        public NeedKind Kind { get; }
        public UtilityCurve Curve { get; }
        public float Weight { get; }

        public NeedTerm(NeedKind kind, UtilityCurve curve, float weight)
        {
            Kind = kind;
            Curve = curve ?? throw new ArgumentNullException(nameof(curve));
            Weight = weight;
        }
    }

    /// <summary>Free-form contextual term (schedule pressure, weather, etc.).</summary>
    public sealed class CustomTerm
    {
        public Func<GoalContext, float> Evaluator { get; }

        public CustomTerm(Func<GoalContext, float> evaluator)
        {
            Evaluator = evaluator ?? throw new ArgumentNullException(nameof(evaluator));
        }
    }

    /// <summary>One additive scoring component of a goal.</summary>
    public sealed class ScoreTerm
    {
        public string Label { get; }
        public float Constant { get; }
        public NeedTerm? Need { get; }
        public TraitTerm? Trait { get; }
        public CustomTerm? Custom { get; }

        private ScoreTerm(string label, float constant, NeedTerm? need, TraitTerm? trait, CustomTerm? custom)
        {
            Label = label;
            Constant = constant;
            Need = need;
            Trait = trait;
            Custom = custom;
        }

        public static ScoreTerm Fixed(string label, float value) => new ScoreTerm(label, value, null, null, null);
        public static ScoreTerm FromNeed(string label, NeedTerm need) => new ScoreTerm(label, 0f, need, null, null);
        public static ScoreTerm FromTrait(string label, TraitTerm trait) => new ScoreTerm(label, 0f, null, trait, null);
        public static ScoreTerm FromCustom(string label, CustomTerm custom) => new ScoreTerm(label, 0f, null, null, custom);
        public static ScoreTerm FromCustom(string label, Func<GoalContext, float> evaluator) => new ScoreTerm(label, 0f, null, null, new CustomTerm(evaluator));
    }

    /// <summary>A selectable goal with explainable scoring configuration.</summary>
    public sealed class GoalDefinition
    {
        public GoalId Id { get; }
        public string DisplayName { get; }
        public float BaseScore { get; }
        public IReadOnlyList<ScoreTerm> Terms { get; }
        /// <summary>Inertia: applied when selecting this goal while another is active.</summary>
        public float SwitchPenalty { get; }
        /// <summary>Needs this goal satisfies when executed.</summary>
        public IReadOnlyList<NeedKind> ReliefNeeds { get; }
        /// <summary>Whether an interrupting need may force-select this goal.</summary>
        public bool CriticalEligible { get; }
        /// <summary>Planner binding: facts that must hold for this goal to be satisfied (Sprint 3).</summary>
        public IReadOnlyList<FactCondition>? DesiredFacts { get; }

        public GoalDefinition(GoalId id, string displayName, float baseScore,
            IEnumerable<ScoreTerm> terms, IEnumerable<NeedKind>? reliefNeeds = null,
            float switchPenalty = 0.08f, bool criticalEligible = false,
            IEnumerable<FactCondition>? desiredFacts = null)
        {
            Id = id;
            DisplayName = displayName ?? throw new ArgumentNullException(nameof(displayName));
            if (!float.IsFinite(baseScore)) throw new ArgumentOutOfRangeException(nameof(baseScore));
            BaseScore = baseScore;
            Terms = terms?.ToList() ?? throw new ArgumentNullException(nameof(terms));
            ReliefNeeds = reliefNeeds?.ToList() ?? new List<NeedKind>();
            SwitchPenalty = switchPenalty >= 0f ? switchPenalty : throw new ArgumentOutOfRangeException(nameof(switchPenalty));
            CriticalEligible = criticalEligible;
            DesiredFacts = desiredFacts?.ToList();
        }
    }

    /// <summary>Snapshot of everything goal scoring may look at.</summary>
    public sealed class GoalContext
    {
        public SimTime Time { get; }
        public NeedSet Needs { get; }
        public PersonalityProfile Personality { get; }
        public GoalId? CurrentGoal { get; }
        /// <summary>0..1 pressure from schedule/job systems (Sprint 5 feeds this).</summary>
        public float SchedulePressure { get; }
        /// <summary>-1..1 mood valence (Sprint 8 feeds this).</summary>
        public float EmotionValence { get; }
        /// <summary>When each goal was last selected (cooldown hysteresis).</summary>
        public IReadOnlyDictionary<GoalId, SimTime> LastSelectedAt { get; }

        public GoalContext(SimTime time, NeedSet needs, PersonalityProfile personality,
            GoalId? currentGoal = null, float schedulePressure = 0f, float emotionValence = 0f,
            IReadOnlyDictionary<GoalId, SimTime>? lastSelectedAt = null)
        {
            Time = time;
            Needs = needs ?? throw new ArgumentNullException(nameof(needs));
            Personality = personality ?? throw new ArgumentNullException(nameof(personality));
            CurrentGoal = currentGoal;
            SchedulePressure = Math.Clamp(schedulePressure, 0f, 1f);
            EmotionValence = Math.Clamp(emotionValence, -1f, 1f);
            LastSelectedAt = lastSelectedAt ?? new Dictionary<GoalId, SimTime>();
        }
    }

    /// <summary>One explained scoring line.</summary>
    public readonly struct ScoreLine
    {
        public string Label { get; }
        public float Value { get; }

        public ScoreLine(string label, float value) { Label = label; Value = value; }

        public override string ToString() =>
            Label + " " + (Value >= 0f ? "+" : "") + Value.ToString("0.00", System.Globalization.CultureInfo.InvariantCulture);
    }

    /// <summary>Final score plus full breakdown for one candidate goal.</summary>
    public sealed class GoalScoreEntry
    {
        public GoalId Goal { get; }
        public string DisplayName { get; }
        public float Final { get; }
        public IReadOnlyList<ScoreLine> Breakdown { get; }
        public bool CriticalOverride { get; }

        internal GoalScoreEntry(GoalId goal, string displayName, float final,
            List<ScoreLine> breakdown, bool criticalOverride)
        {
            Goal = goal;
            DisplayName = displayName;
            Final = final;
            Breakdown = breakdown;
            CriticalOverride = criticalOverride;
        }
    }

    /// <summary>Ranked outcome of goal selection.</summary>
    public sealed class GoalSelectionResult
    {
        public GoalScoreEntry? Winner { get; }
        public IReadOnlyList<GoalScoreEntry> Ranked { get; }
        /// <summary>True when an existing commitment beat the raw winner (hysteresis).</summary>
        public bool KeptPreviousGoal { get; }

        internal GoalSelectionResult(GoalScoreEntry? winner, List<GoalScoreEntry> ranked, bool keptPrevious)
        {
            Winner = winner;
            Ranked = ranked;
            KeptPreviousGoal = keptPrevious;
        }
    }
}

File: packages/cognition/src/goals/goalModel.ts
/**
 * Goal model: explainable additive scoring terms, switch inertia, re-selection
 * cooldown, critical-need override. Deterministic: score desc, then goal id ordinal.
 */
import type { NeedKind, NeedSet, NeedState } from "../needs/needs.js";
import type { PersonalityProfile, PersonalityTrait } from "../personality/personality.js";
import type { UtilityCurve } from "../utility/curves.js";
import type { FactCondition } from "../goap/plannerState.js";

export enum WeatherState {
  Clear = 0,
  Cloudy = 1,
  Rain = 2,
  HeavyRain = 3,
}

export interface PreferenceProfileLike {
  get(key: string): number;
}
export const EMPTY_PREFERENCES: PreferenceProfileLike = { get: () => 0 };
export interface GoalContext {
  timeMinutes: number;
  needs: NeedSet;
  personality: PersonalityProfile;
  currentGoal?: string | undefined;
  schedulePressure: number; // 0..1
  emotionValence: number;   // -1..1
  lastSelectedAt?: ReadonlyMap<string, number>;
  weather: WeatherState;
  preferences: PreferenceProfileLike;
  /** Location key of the resident's current location ("loc_cafe"), if placed. */
  currentLocationKey?: string;
  /** Bounded habit bonus (≤0.15) supplied by the host for the current spot. */
  habitBonus?: number;
  /** Bounded long-term-intention bias (−0.1..+0.1), host-derived. */
  intentionBias?: number;
  /** Occupancy ratio (0..1+) of the resident's current venue, if any. */
  crowding?: number;
}

export interface ScoreLine { label: string; value: number }

export class GoalScoreEntry {
  constructor(
    readonly goal: string,
    readonly displayName: string,
    readonly final: number,
    readonly breakdown: readonly ScoreLine[],
    readonly criticalOverride: boolean,
  ) {}
}

export class GoalSelectionResult {
  constructor(
    readonly winner: GoalScoreEntry | null,
    readonly ranked: readonly GoalScoreEntry[],
  ) {}
}

export interface NeedTerm { kind: NeedKind; curve: UtilityCurve; weight: number }
export interface TraitTerm { trait: PersonalityTrait; min: number; max: number }

export type ScoreTerm =
  | { kind: "fixed"; label: string; value: number }
  | { kind: "need"; label: string; term: NeedTerm }
  | { kind: "trait"; label: string; term: TraitTerm }
  | { kind: "custom"; label: string; evaluate: (ctx: GoalContext) => number };

export const fixedTerm = (label: string, value: number): ScoreTerm => ({ kind: "fixed", label, value });
export const needTerm = (label: string, t: NeedTerm): ScoreTerm => ({ kind: "need", label, term: t });
export const traitTerm = (label: string, t: TraitTerm): ScoreTerm => ({ kind: "trait", label, term: t });
export const customTerm = (label: string, fn: (ctx: GoalContext) => number): ScoreTerm => ({ kind: "custom", label, evaluate: fn });

export interface GoalDefinition {
  id: string;
  displayName: string;
  baseScore: number;
  terms: readonly ScoreTerm[];
  reliefNeeds: readonly NeedKind[];
  switchPenalty: number;
  criticalEligible: boolean;
  desiredFacts?: readonly FactCondition[] | undefined;
}

export function defineGoal(
  id: string, displayName: string, baseScore: number, terms: readonly ScoreTerm[],
  opts: {
    reliefNeeds?: readonly NeedKind[];
    switchPenalty?: number;
    criticalEligible?: boolean;
    desiredFacts?: readonly FactCondition[];
  } = {},
): GoalDefinition {
  if (!Number.isFinite(baseScore)) throw new Error("baseScore must be finite");
  return {
    id, displayName, baseScore, terms,
    reliefNeeds: opts.reliefNeeds ?? [],
    switchPenalty: opts.switchPenalty ?? 0.08,
    criticalEligible: opts.criticalEligible ?? false,
    desiredFacts: opts.desiredFacts,
  };
}

export interface GoalSelectionOptions {
  criticalBonus: number;
  reSelectionCooldownMinutes: number;
  reSelectionPenalty: number;
}

export const DEFAULT_SELECTION_OPTIONS: GoalSelectionOptions = {
  criticalBonus: 0.5,
  reSelectionCooldownMinutes: 45,
  reSelectionPenalty: 0.12,
};

function containsKind(list: readonly NeedKind[], kind: NeedKind): boolean {
  return list.includes(kind);
}

export function selectGoals(
  ctx: GoalContext,
  candidates: readonly GoalDefinition[],
  options: GoalSelectionOptions = DEFAULT_SELECTION_OPTIONS,
): GoalSelectionResult {
  const interrupting: NeedState | null = ctx.needs.findInterrupting();
  const ranked: GoalScoreEntry[] = [];

  for (const goal of candidates)
    ranked.push(scoreGoal(ctx, goal, interrupting, options));

  ranked.sort((x, y) => {
    const byScore = y.final - x.final;
    return byScore !== 0 ? byScore : x.goal < y.goal ? -1 : x.goal > y.goal ? 1 : 0;
  });

  return new GoalSelectionResult(candidates.length > 0 ? ranked[0]! : null, ranked);
}

function scoreGoal(
  ctx: GoalContext, goal: GoalDefinition,
  interrupting: NeedState | null, options: GoalSelectionOptions,
): GoalScoreEntry {
  const lines: ScoreLine[] = [];
  const isCurrent = ctx.currentGoal !== undefined && ctx.currentGoal === goal.id;
  let criticalOverride = false;
  let total = goal.baseScore;

  lines.push({ label: "Base", value: goal.baseScore });

  for (const term of goal.terms) {
    let value = 0;
    switch (term.kind) {
      case "fixed": value = term.value; break;
      case "need":
        value = term.term.curve.evaluate(ctx.needs.get(term.term.kind).normalized) * term.term.weight;
        break;
      case "trait": {
        const v = ctx.personality.get(term.term.trait);
        value = term.term.min + (term.term.max - term.term.min) * v;
        break;
      }
      case "custom":
        value = term.evaluate(ctx);
        if (!Number.isFinite(value))
          throw new Error(`Custom term '${term.label}' on '${goal.id}' produced a non-finite value.`);
        break;
    }
    total += value;
    lines.push({ label: term.label, value });
  }

  if (interrupting !== null && goal.criticalEligible &&
      containsKind(goal.reliefNeeds, interrupting.definition.kind)) {
    total += options.criticalBonus;
    lines.push({ label: "CriticalOverride", value: options.criticalBonus });
    criticalOverride = true;
  } else {
    if (!isCurrent && ctx.currentGoal !== undefined && goal.switchPenalty > 0) {
      total -= goal.switchPenalty;
      lines.push({ label: "SwitchCost", value: -goal.switchPenalty });
    }
    if (!isCurrent && ctx.lastSelectedAt) {
      const last = ctx.lastSelectedAt.get(goal.id);
      if (last !== undefined && ctx.timeMinutes - last < options.reSelectionCooldownMinutes) {
        total -= options.reSelectionPenalty;
        lines.push({ label: "RecentCooldown", value: -options.reSelectionPenalty });
      }
    }
  }

  if (!Number.isFinite(total)) throw new Error(`Goal '${goal.id}' scored non-finite.`);
  return new GoalScoreEntry(goal.id, goal.displayName, total, lines, criticalOverride);
}
